### PR TITLE
EncoderConfig and AtomicLevel validation added

### LINKED ISSUE #777: Add Config validation to build stage to prevent runtime panics
When creating a custom zap.Config there seems to be only minimal validation of the configuration that results in runtime panics if certain properties are missing. This is dangerous as all required configuration properties should be checked at the build stage. 

There is probably a reason this has not been done, but opening an issue nonetheless.

## Example

```
	cfg := zap.Config{
		Encoding:    "json",
		OutputPaths: []string{"stderr"},
		EncoderConfig: zapcore.EncoderConfig{
			MessageKey: "msg",
		},
	}

	logger, err := cfg.Build()
	if err != nil {
		panic(err)
	}

	logger.Sugar().Errorw("Problem here", "foo", "bar") // Panic
```

This results in the following panic:

```

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x12696c6]

goroutine 1 [running]:
go.uber.org/atomic.(*Int32).Load(...)
	/Users/owner/go/pkg/mod/go.uber.org/atomic@v1.5.1/atomic.go:41
go.uber.org/zap.AtomicLevel.Level(...)
	/Users/owner/go/pkg/mod/go.uber.org/zap@v1.13.0/level.go:97
go.uber.org/zap.AtomicLevel.Enabled(0x0, 0x10b5c02, 0xc000098c30)
	/Users/owner/go/pkg/mod/go.uber.org/zap@v1.13.0/level.go:92 +0x6
go.uber.org/zap.(*SugaredLogger).log(0xc000055f00, 0xc000010002, 0x1312a3e, 0xc, 0x0, 0x0, 0x0, 0xc000055f30, 0x2, 0x2)
	/Users/owner/go/pkg/mod/go.uber.org/zap@v1.13.0/sugar.go:221 +0x1aa
go.uber.org/zap.(*SugaredLogger).Errorw(...)
	/Users/owner/go/pkg/mod/go.uber.org/zap@v1.13.0/sugar.go:191
main.main()
	/Users/owner/zaplog/zaplog.go:22 +0x27b
exit status 2
```

The issue here is we are missing a "Level"

```
Level:            zap.NewAtomicLevelAt(zapcore.DebugLevel),
```

However, there are other issues as well. Suppose you specify the "TimeKey", but not the "EncodeTime".

```
	cfg := zap.Config{
		Level:       zap.NewAtomicLevelAt(zapcore.DebugLevel),
		Encoding:    "json",
		OutputPaths: []string{"stderr"},
		EncoderConfig: zapcore.EncoderConfig{
			MessageKey: "msg",
			TimeKey:    "ts",
		},
	}
```

This configuration will build without error, but you again will be faced with `runtime error: invalid memory address or nil pointer dereference` on use.

### PR DESCRIPTION
**Pull request by YashishDua**
Resolves #777

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -174,6 +174,13 @@ func (cfg Config) Build(opts ...Option) (*Logger, error) {
 		return nil, err
 	}
 
+	if cfg.Level == (AtomicLevel{}) {
```
Comment (abhinav) [inline config.go]: The parentheses are unnecessary.

```suggestion
	if cfg.Level == AtomicLevel{} {
```

File: config.go [hunk 2]
```diff
@@ -174,6 +174,13 @@ func (cfg Config) Build(opts ...Option) (*Logger, error) {
...
+		cfg.Level = NewAtomicLevelAt(InfoLevel)
+		if cfg.Development {
+			cfg.Level = NewAtomicLevelAt(DebugLevel)
```
Comment (abhinav) [inline config.go]: Should we return an `error` instead?

If we prefer to add a default behavior for Level when unspecified, we'd probably want to document it on Config.Level.

File: config.go [hunk 3]
```diff
@@ -239,5 +246,12 @@ func (cfg Config) openSinks() (zapcore.WriteSyncer, zapcore.WriteSyncer, error)
 }
 
 func (cfg Config) buildEncoder() (zapcore.Encoder, error) {
+	if cfg.EncoderConfig.TimeKey != "" && cfg.EncoderConfig.EncodeTime == nil {
+		cfg.EncoderConfig.EncodeTime = zapcore.EpochTimeEncoder
```
Comment (abhinav) [inline config.go]: Same here. It's not obvious why we pick epoch for production and ISO8601 for development. Perhaps an error would be preferable?

**Comment by prashantv** [comment]
Agree with @abhinav -- I think `Development` changing time behaviour might be surprising, since the current docs for `Development` is documented only to put the logger in development mode.

```
// Development puts the logger in development mode, which changes the
// behavior of DPanicLevel and takes stacktraces more liberally.
```

To avoid any surprises, returning an error seems best. We should also add unit tests to validate the new behaviour. Thanks!

File: config.go (hunk 2 above)
Comment (YashishDua) [reply config.go]: I could see InfoLevel as default in ProductionConfig and DebugLevel in DevelopmentConfig, that's why thought of adding the same defaults here.

Let's return an error then!

File: config.go (hunk 3 above)
Comment (YashishDua) [reply config.go]: NewDevelopmentEncoderConfig specifies ISO8601TimeEncoder by default. I have not made any explicit choices here. But, let's return an error.

File: config.go (hunk 1 above)
Comment (YashishDua) [reply config.go]: Tried this earlier, but comes with `expected ';', found '{'`

File: config_test.go [hunk 4]
```diff
@@ -106,3 +107,39 @@ func TestConfigWithInvalidPaths(t *testing.T) {
...
+			},
+			expectErr: "missing EncodeTime in EncoderConfig",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			cfg := tt.cfg
+			_, err := cfg.Build()
+			assert.Equal(t, err.Error(), tt.expectErr)
```
Comment (abhinav) [inline config_test.go]: Testify assertions are in the order: expected, then actual.

```suggestion
			assert.Equal(t, tt.expectErr, err.Error())
```

File: config_test.go [hunk 5]
```diff
@@ -106,3 +107,39 @@ func TestConfigWithInvalidPaths(t *testing.T) {
...
+				},
+			},
+			expectErr: "missing EncodeTime in EncoderConfig",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			cfg := tt.cfg
+			_, err := cfg.Build()
```
Comment (abhinav) [inline config_test.go]: err.Error on the following line will panic if we don't first verify that the error is non-nil.
We should use require.Error here.

```suggestion
			_, err := cfg.Build()
			require.Error(t, err)
```

File: config_test.go [hunk 6]
```diff
@@ -106,3 +107,39 @@ func TestConfigWithInvalidPaths(t *testing.T) {
...
 }
+
+func TestConfigWithMissingAttributes(t *testing.T) {
+	tests := []struct {
+		desc      string
+		cfg       Config
+		expectErr string
+	}{
+		{
+			desc: "missing_level",
```
Comment (abhinav) [inline config_test.go]: We don't need to add `_` to subtest case names here. testing.T will do it for us.

```suggestion
			desc: "missing level",
```

File: config_test.go [hunk 7]
```diff
@@ -106,3 +107,39 @@ func TestConfigWithInvalidPaths(t *testing.T) {
...
+			cfg: Config{
+				Encoding: "json",
+			},
+			expectErr: "missing Level",
+		},
+		{
+			desc: "missing_encoder_time_in_encoder_config",
```
Comment (abhinav) [inline config_test.go]:
```suggestion
			desc: "missing encoder time in encoder config",
```

File: config.go [hunk 8]
```diff
@@ -239,5 +244,9 @@ func (cfg Config) openSinks() (zapcore.WriteSyncer, zapcore.WriteSyncer, error)
 }
 
 func (cfg Config) buildEncoder() (zapcore.Encoder, error) {
+	if cfg.EncoderConfig.TimeKey != "" && cfg.EncoderConfig.EncodeTime == nil {
```
Comment (abhinav) [inline config.go]: It seems that newEncoder is the ultimate constructor for zapcore.Encoder given
an EncoderConfig. We should maybe move this check to the newEncoder function,
leaving buildEncoder unchanged.

**Pull request merged**